### PR TITLE
Refactor the leadership module to make use of existing and new events

### PR DESCRIPTION
**Pull request by wankdanker**
It is not totally necessary to create a whole new class/interface to
manage the leadership. With the addition of a few new events, it's
possible to wire everything up just using events. I feel like this
makes it more generic as well.

This commit removes most of the internal references to leadershipElector
and modifies the BasicLeadershipElection class to accept the node-discover
instance in the constructor. The constructor then wires all the events to
the class methods.

This restores the expected behavior that is reported in #36

ref: #36, #35, #34

@aikar Would this still work with what you needed?

### CODE REVIEW

**Comment by aikar** [comment]
My desire was to completely replace the implementation. I've cloned the BasicLeadership class and modified it to suit our needs (in our specific case, we have a separate master per "application type", so multiple types of applications are part of a larger network that one type can talk to another type, but within the sub types each can have a master.

this is done in our case by the check method skipping evaluating nodes of another type.

So I'm not sure this style would enable my goal of completely replacing the master system (I have other ideas for a even more complex leadership pattern)

Replacing an implementation is a safer route in my opinion, where events simply only allow reactions. I want to control when and what events fire.

I have separate questions on #36 ill ask there.

**Comment by aikar** [comment]
I originally considered going this route too, but ended up never doing it but agree it is a good thing.

However I wanted to use start as a hook point to say "intercept starting, allowing preparing the instance before continuing".

For example, I hooked in my own extra properties on discover.me before starting so the first payload can have it.

My concern is that the discover instance auto starts, and I wanted a way to control that. 

I guess alternatively, autostart:false as an option and allow calling .start() manually?

**Comment by aikar** [comment]
Also I went with the original start route and pass my own instance as I wanted to use dependency injection on my leadership elector instance, so that it has references to my own systems objects, which timing wise is difficult with the current auto start.

atm I have leadershipElector: new MyLeadershipElector(this) then set this.discover in .start() so I have access to both guaranteed at any event fire.

**Comment by aikar** [comment]
Could solve that by moving the event binders to the start method and keep it like I originally had it with constructor-less invocation.

Doing it that way I think also avoids making this an API breaking change (considering the new .start() API is now released and I'm using it in my own code, plus passing an instantiated object vs the constructor in options)

**Comment by wankdanker** [comment]
What if there was just a way to turn off the internal promotion/demotion code (which is now a module)? That in addition to the `start : false` option, you could just instantiate it and then wire up all the events to do whatever you want. I think that would have been the ideal way to go.

But, since the API has already been released, I can add some extra logic depending on the value of `opts.leadershipElector`.

- if `opts.leadershipElector` is a function
  - instantiate it with self as the constructor
  - bind events to methods
  - also call start method passing in self?
- if `opts.leadershipElector` is an object
  - bind events to methods
  - call start method passing in self
- if `opts.leadershipElector` is null or undefined
  - use the backwards compatible leadership module
- if `opts.leadershipElector` === false
  - don't do any leadership stuff

File: lib/leadership.js [hunk 1]
```diff
@@ -1,22 +1,66 @@
...
+ * @param {*} discover the discover instance from which events will be bound to LeadershipElection instance methods
+ * @returns {LeadershipElection} the instance of the LeadershipElection module that was resolved or false
+ */
+function resolveLeadership (leadershipElector, discover) {
+	let elector;
+
+	if (leadershipElector === false) {
+		elector = false;
+	}
+	else if (leadershipElector == null || leadershipElector == undefined) {
```
Comment (aikar) [inline lib/leadership.js]: == null covers undefined already

File: lib/discover.js [hunk 2]
```diff
@@ -217,31 +228,27 @@ function Discover (options, callback) {
 
 			return false;
 		}
-
-		self.leadershipElector.start(self, function (err) {
+		
+		self.broadcast.start(function (err) {
```
Comment (aikar) [inline lib/discover.js]: Need to keep this section as it was, as in my use case where I passed my own instantiation, I have no reference to the discover object yet.
start was my hook into receiving it.

auto connect false would open door to inject it myself, but say I instantiate BasicLeadership myself and extend it, that code depends on having a reference to this.discover.

That needs to be maintained.

Otherwise the resolveLeadership needs to inject the discover reference.

I think if you revert the start changes here, so that .start() can still be used to "intercept" start, then this would not be an API breaking change and overall looks good to me.

File: lib/leadership.js [hunk 3]
```diff
@@ -1,22 +1,66 @@
+module.exports = resolveLeadership;
```
Comment (aikar) [inline lib/leadership.js]: not really a fan of this style, my preference is to simply export symbols only, so that

import {resolveLeadership, BasicLeadershipElection} from "node-discover/leadership" 

works

With this style, its 

import * as resolveLeadership, {BasicLeadershipElection} from "node-discover/leadership" 

which is not clean IMO.

recommended to go back to module.exports = { ... }

File: lib/leadership.js [hunk 4]
```diff
@@ -82,16 +122,8 @@ BasicLeadershipElection.prototype.check = function () {
 	}
 };
 
-BasicLeadershipElection.prototype.start = function (discover, callback) {
-	this.discover = discover;
```
Comment (aikar) [inline lib/leadership.js]: Per other review comment, I believe this should be kept.

File: examples/master.js [hunk 5]
```diff
@@ -10,28 +10,32 @@
...
+	key : process.argv[2], weight : Date.now() * -1
+	, mastersRequired : 2
+	//, leadershipElector : Discover.NoLeadershipElection
+});
 
 console.log('I am ' + d.broadcast.instanceUuid);
 
 d.on("added", function (obj) {
-	console.log("New node discovered on the network.");
+	process.stdout.write("\nNew node discovered on the network.");
```
Comment (aikar) [inline examples/master.js]: Why was these changed? I know it doesn't impact the library, just curious.

File: lib/leadership.js (hunk 1 above)
Comment (wankdanker) [reply lib/leadership.js]: Fair enough.

File: lib/discover.js (hunk 2 above)
Comment (wankdanker) [reply lib/discover.js]: There is a `started` event that is emitted which is bound to the start method of the Leadership instance which receives the instance of node-discover. So, unless I'm misunderstanding, I think this is covered without having to revert this change.

File: lib/leadership.js (hunk 4 above)
Comment (wankdanker) [reply lib/leadership.js]: Understood, if you are inheriting from this, then that makes sense to leave it. It is not necessary though for the internal default use of the module.

File: lib/leadership.js (hunk 3 above)
Comment (wankdanker) [reply lib/leadership.js]: Are you using resolveLeadership from outside the module?

File: examples/master.js (hunk 5 above)
Comment (wankdanker) [reply examples/master.js]: I use this example to test the behavior. The output bothered me because there were too many blank lines and it didn't feel like a repl.

File: lib/leadership.js (hunk 3 above)
Comment (wankdanker) [reply lib/leadership.js]: Also, I exposed the default leadership modules on the node-discover constructor.

Can't you do:

```js
import { BasicLeadershipElection } from 'node-discover'
```

File: lib/discover.js (hunk 2 above)
Comment (aikar) [reply lib/discover.js]: slight behavioral change in that it no longer allows controlling the start, but I'm ok with that as i can refactor my code to account for it, since the start param can control delaying start too.

File: lib/leadership.js (hunk 3 above)
Comment (aikar) [reply lib/leadership.js]: I wasn't meaning as part of the public api, this wouldn't impact me, just my opinion on defining exports, I don't see value in this vs just exporting resolve as a symbol.

typically default exports like this is used for classes when the entire point of the file is the class. 

i'm in favor of dealing with exports in a way that is cleanly importable in module format.

Comment (aikar) [reply lib/leadership.js]: Also it's simpler for type definement too ie for typescript .d.ts files (this syntax is possible yeah, just not as 'clean' IMO)

**Comment by aikar** [comment]
Left some feedback but I think overall this is good w/ me now. the .start change is technically an API breaking change as it changes existing behavior (and I will have to update code to account for that) but I'm ok with that since the start option was added.

If you want to skip bumping semver major, I would be ok with that as i'm pretty confident i'm the only one this would impact and i'll resolve it on next update...

File: lib/discover.js (hunk 2 above)
Comment (wankdanker) [reply lib/discover.js]: Oh I think I understand now. You were controlling the start by waiting to call the callback that was passed to the start method. I can't think of a clever way to handle this without reverting the change. Thank you for refactoring your code to account for the change.

File: lib/leadership.js (hunk 3 above)
Comment (wankdanker) [reply lib/leadership.js]: Thank you for your comments on this. I will take your feedback under advisement.

**Comment by wankdanker** [comment]
> If you want to skip bumping semver major, I would be ok with that as i'm pretty confident i'm the only one this would impact and i'll resolve it on next update...

Cool, thank you. I'll merge this and release 1.1.2.

**Comment by wankdanker** [comment]
node-discover@1.1.2 is on npm.
